packages/plugins/documentation/__tests__/build-component-schema.test.js

```diff
@@ -0,0 +1,321 @@
+'use strict';
+
+const buildComponentSchema = require('../server/services/helpers/build-component-schema');
+const strapi = require('../__mocks__/strapi');
+
+describe('Build Component Schema', () => {
+  it('builds the Response schema', () => {
+    global.strapi = strapi;
+
+    const apiMocks = [
+      {
+        name: 'users-permissions',
+        getter: 'plugin',
+        ctNames: ['role'],
+      },
+      { name: 'restaurant', getter: 'api', ctNames: ['restaurant'] },
+    ];
+
+    let schemas;
+    for (const mock of apiMocks) {
+      schemas = {
+        ...schemas,
+        ...buildComponentSchema(mock),
+      };
+    }
+
+    const schemaLength = Object.keys(schemas).length;
+    const [pluginResponseName, apiResponseName] = Object.keys(schemas);
+    const [pluginResponseValue, apiResponseValue] = Object.values(schemas);
+
+    const expectedShape = {
+      properties: {
+        data: {
+          type: 'object',
+          properties: {
+            id: { type: 'string' },
+            attributes: { type: 'object', properties: { test: { type: 'string' } } },
+          },
+        },
+        meta: { type: 'object' },
+      },
+    };
+
+    expect(schemaLength).toBe(2);
+    expect(pluginResponseName).toBe('UsersPermissionsRoleResponse');
+    expect(apiResponseName).toBe('RestaurantResponse');
+    expect(pluginResponseValue).toStrictEqual(expectedShape);
+    expect(apiResponseValue).toStrictEqual(expectedShape);
+  });
+
+  it('builds the ResponseList schema', () => {
+    global.strapi = {
```

Comment:
Since I believe only the routes change, maybe you can mutate that key directly. Something like:

```js
global.strapi = { ...strapi } // or Object.assign({}, strapi)
global.strapi.plugins['users-permissions'].routes['content-api'].routes = [/* items */]
```

and same for the restaurant routes. It may be a personal preference but I think it's easier to spot what actually changes instead of rebuilding the object structure

Reply:
Yes that will be cleaner and more readable. I had to use lodash cloneDeep since spread and Object.assign create only a shallow copy of an object and nested objects seem to still reference / mutate their origin.

Reply:
Indeed I hadn't thought about that